src/QuadSort.jl

```diff
@@ -0,0 +1,951 @@
+        parity_merge_four!(array, array_index, swap, swap_index, cmp)
+        parity_merge_four!(array, array_index +8, swap, swap_index +8, cmp)
+        parity_merge!(array, array_index, swap, swap_index, UInt(8), UInt(8), cmp)
+    end
+end
+
+function tail_swap!(array, array_index::Int, swap, swap_index::Int, nmemb::UInt, cmp)
+    @inbounds begin
+        if nmemb < 5
+            tiny_sort!(array, array_index, nmemb, cmp)
+            return
+        end
+        if nmemb < 8
+            quad_swap_four!(array, array_index, cmp)
+            twice_unguarded_insert!(array, array_index, UInt(4), nmemb, cmp)
+            return
+        end
+        if nmemb < 12
+            parity_swap_eight!(array, array_index, swap, swap_index, cmp)
+            twice_unguarded_insert!(array, array_index, UInt(8), nmemb, cmp)
+            return
+        end
+        if 16 ≤ nmemb < 24
+            parity_swap_sixteen!(array, array_index, swap, swap_index, cmp)
+            twice_unguarded_insert!(array, array_index, UInt(16), nmemb, cmp)
+            return
+        end
+        half1 = asInt(nmemb ÷ 2)
+        quad1 = half1 ÷ 2
+        quad2 = half1 - quad1
+
+        half2 = asInt(nmemb - half1)
+        quad3 = half2 ÷ 2
+        quad4 = half2 - quad3
+
+        pta = array_index
+        tail_swap!(array, pta, swap, swap_index, asUInt(quad1), cmp); pta += quad1
+        tail_swap!(array, pta, swap, swap_index, asUInt(quad2), cmp); pta += quad2
+        tail_swap!(array, pta, swap, swap_index, asUInt(quad3), cmp); pta += quad3
+        tail_swap!(array, pta, swap, swap_index, asUInt(quad4), cmp)
+
+        cmp(array[array_index+quad1-1], array[array_index+quad1]) ≤ 0 &&
+            cmp(array[array_index+half1-1], array[array_index+half1]) ≤ 0 &&
+            cmp(array[pta-1], array[pta]) ≤ 0 && return
+
+        parity_merge!(swap, swap_index, array, array_index, asUInt(quad1), asUInt(quad2), cmp)
+        parity_merge!(swap, swap_index + half1, array, array_index + half1, asUInt(quad3), asUInt(quad4), cmp)
+        parity_merge!(array, array_index, swap, swap_index, asUInt(half1), asUInt(half2), cmp)
+    end
+end
+
+# the next three functions create sorted blocks of 32 elements
+function quad_reversal!(array, pta::Int, ptz::Int)
+    @inbounds begin
+        loop = (ptz - pta) ÷ 2
+
+        ptb = pta + loop
+        pty = ptz - loop
+        if iseven(loop)
+            array[ptb], array[pty] = array[pty], array[ptb]
+            ptb -= 1; pty += 1
+            loop -= 1
+        end
+        loop ÷= 2
+        while true
+            array[pta], array[ptz] = array[ptz], array[pta]
+            pta += 1; ptz -= 1
+            array[ptb], array[pty] = array[pty], array[ptb]
+            ptb -= 1; pty += 1
+
+            iszero(loop) && break
+            loop -= 1
+        end
+        nothing
+    end
+end
+
+function quad_swap_merge!(array, array_index::Int, swap, swap_index::Int, cmp)
```

Comment:
```suggestion
function quad_swap_merge!(array, array_index::Int, swap, swap_index::Int, cmp::T) where T # Force specialization on cpm
```
This forces specialization on cmp which, by default, is not specialized on because (I'm guessing) cmp is a function. (see https://docs.julialang.org/en/v1/manual/performance-tips/#Be-aware-of-when-Julia-avoids-specializing). The `parity_merge_two` function calls `cmp` so it specializes, hence the call from quad_swap_merge to parity_merge_two uses dynamic dispatch.

```julia
julia> @b rand(100_000) sort!(_, alg=SortingAlgorithms.QuadSort) evals=1 # Before
4.704 ms (76922 allocs: 2.525 MiB)

shell> git stash pop
On branch blitsort
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   src/QuadSort.jl

no changes added to commit (use "git add" and/or "git commit -a")
Dropped refs/stash@{0} (572b55ea506e7bd86da6622f0ffe56919a6c0115)

julia> @b rand(100_000) sort!(_, alg=SortingAlgorithms.QuadSort) evals=1 # Suggestion
3.857 ms (2126 allocs: 830.938 KiB)
```